bestie/server/xml_result.go

```diff
@@ -0,0 +1,279 @@
+	for _, ts := range testSuites.TestSuites {
+		// Check for optional BigQuery table specification within testsuite XML data.
+		dataset := ""
+		tableName := ""
+		for _, prop := range ts.Properties.Props {
+			if prop.Name == "bq_dataset" {
+				dataset = prop.Value
+			}
+			if prop.Name == "bq_tablename" {
+				tableName = prop.Value
+			}
+		}
+		result.table = bigQueryTable{
+			// The project field is omitted from the XML data.
+			dataset:   dataset,
+			tableName: tableName,
+		}
+
+		// The presence of a <system-out> tag means the output results are unformatted and
+		// must be processed by scraping the console output to the collect test case info.
+		var xmlResults []*xmlResult
+		var xerr error
+		if len(ts.SystemOut) > 0 {
+			xmlResults, xerr = parseUnstructuredXml(&ts)
+		} else {
+			xmlResults, xerr = parseStructuredXml(&ts)
+		}
+		if xerr != nil {
+			return nil, fmt.Errorf("Error parsing XML test results: %w", xerr)
+		}
+
+		for _, xr := range xmlResults {
+			// Construct the BigQuery metric from the XML results provided.
+			var m testMetric = testMetric{
+				metricName: "testresult",
+				tags: map[string]string{
+					"result":     xr.result,
+					"test_case":  xr.tcTestCase,
+					"test_class": xr.tcClass,
+					"test_file":  xr.tcFile,
+					"type":       "summary",
+				},
+				value:     xr.duration,
+				timestamp: xr.testTime.UnixNano(),
+			}
+			// Append any test case properties to result tags.
+			for k, v := range xr.tcProps {
+				m.tags[k] = v
+			}
+			result.metrics = append(result.metrics, m)
+		}
+	}
+	return &result, nil
+}
+
+// Parse a test.xml file that is output in a "raw" format, where individual
+// test case names and pass/fail results must be scraped from the console
+// output text string. This is the default format produced by 'bazel test'.
+func parseUnstructuredXml(ts *TestSuite) ([]*xmlResult, error) {
+	// Read each line of the console string, looking for two specific lines for each test case:
+	//
+	//   systest/trial/test_app.py::TestHello::test_app_metrics_nofile Testing resource 1322
+	//   MAKEREPORT: test_app.py::TestHello::test_app_metrics_nofile: passed
+	//
+	// The last token in the MAKEREPORT: line is one of: passed, failed, or skipped.
+	//
+	// An abbreviated form of a skipped test is reported as follows:
+	//
+	//   systest/trial/test_app.py::TestHello::test_app_metrics_nofile SKIPPED
+	//
+	// Also look for a single line containing the SKIPPEDSESSIONFINISH: token designating
+	// a test case that contains a pytest.skip() call during its execution:
+	//
+	//   systest/trial/test_app.py::TestHello::test_app_metrics_file[test.metrics.pb] SKIPPEDSESSIONFINISH: ...
+	//
+	// Extract the full test case ID from the first entry and the passed/failed/skipped result
+	// from the second.
+	matches := make(map[string]string)
+	var key string
+	for _, line := range strings.Split(strings.TrimSuffix(ts.SystemOut, "\n"), "\n") {
```

Comment:
strings.TrimSpace instead of strings.TrimSuffix

Reply:
Fixed.